src/utils/daos_hdlr.c

```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+				D_FREE(buf);
+			daos_file_close(daos_src_file, filename);
+			daos_file_close(daos_dst_file, filename);
+		} else if (S_ISDIR(st.st_mode)) {
+			/* Check that the directory is not "d" or d's parent. */
+			if ((strcmp(d_name, "..") != 0) && (strcmp(d_name, ".") != 0)) {
+				char path[MAX_FILENAME];
+				char dpath[MAX_FILENAME];
+				path_length = snprintf(path, MAX_FILENAME, "%s", filename);
+				if (daos_src_file->type == DAOS && daos_dst_file->type == POSIX)
+					path_length = snprintf(dpath, MAX_FILENAME, "%s", dst_filename);
+				if (path_length >= MAX_FILENAME) 
+					fprintf(stderr, "Path length is too long.\n");
+				/* TODO create dst with write permissions, then after copying
+		 		 * make sure to change permissions to match the source */
+				if (daos_src_file->type == DAOS && daos_dst_file->type == POSIX) {
+					rc = daos_file_mkdir(daos_dst_file, dpath, st.st_mode);
+				} else {
+					rc = daos_file_mkdir(daos_dst_file, path, st.st_mode);
+				}
+				/* Recursively call "fs_copy" with the new path. */
+				rc = fs_copy(daos_src_file, daos_dst_file, path, fs_dst_prefix);
+			}
+		} else {
+			rc = ENOTSUP;
+			fprintf(stderr, "file type is not supported (%d)\n", rc);
+			D_GOTO(out, rc);
+		}
+	}
+
+	/* After going through all the entries, close the directory. */
+	if (daos_file_closedir(daos_src_file, src_dir))
+		fprintf(stderr, "Could not close '%s': %s\n",
+			dir_name, strerror (errno));
+out:
+	return rc;
+}
+
+static int
+fs_copy_connect(daos_file_t *daos_src_file,
+		   daos_file_t *daos_dst_file,
+		   struct cmd_args_s *ap,
+		   daos_cont_info_t *src_cont_info,
+		   daos_cont_info_t *dst_cont_info,
+		   struct duns_attr_t *src_dattr,
+		   struct duns_attr_t *dst_dattr)
+{
+	/* connect to source pool/conts */
+	int rc = 0;
+	if (daos_uuid_valid(ap->src_p_uuid)) { 
+		rc = daos_pool_connect(ap->src_p_uuid, ap->sysname, NULL,
+				DAOS_PC_RW, &ap->pool, NULL, NULL);
+		if (rc != 0) {
+			fprintf(stderr, "failed to connect to destination pool: %d\n", rc);
+			/* if connection to source fails, do not continue */
+			D_GOTO(out, rc);
+		} else {
+			if (daos_uuid_valid(ap->src_c_uuid)) { 
+				rc = daos_cont_open(ap->pool, ap->src_c_uuid, DAOS_COO_RW,
+					&ap->cont, src_cont_info, NULL);
+				if (rc != 0) {
+					fprintf(stderr, "failed to open source cont: %d\n", rc);
+					D_GOTO(out, rc);
+				}
+				rc = dfs_mount(ap->pool, ap->cont, O_RDWR, &daos_src_file->dfs);
+				if (rc) {
+					fprintf(stderr, "dfs mount on source failed: %s, %d\n",
+						ap->src_path, rc);
+					D_GOTO(out, rc);
+				}
+			}
+		}
+	} else {
+		/* Resolve pool, container UUIDs from src path if needed */
+        	if (ap->src_path != NULL) {
+			rc = duns_resolve_path(ap->src_path, src_dattr);
+			if (rc == 0) { 
+				ap->type = (*src_dattr).da_type;
+				uuid_copy(ap->src_p_uuid, (*src_dattr).da_puuid);
+				uuid_copy(ap->src_c_uuid, (*src_dattr).da_cuuid);
+				rc = daos_pool_connect(ap->src_p_uuid, ap->sysname, NULL,
```

Comment:
seems like a duplicate of the connect / open/ mount above.
to avoid duplication we can do the checks and resolve the pool and cont uuid from path or input paramters then do a single point where we do connect / open /mount

Reply:
@mchaarawi added